rfd/0175-spiffe-federation.md

```diff
@@ -0,0 +1,429 @@
+---
+author: Noah Stride (noah@goteleport.com)
+state: draft
+---
+
+# RFD 175 - SPIFFE Federation and SPIFFE Join Method
+
+## Required Approvers
+
+- Engineering: @timothyb89
+- Product: @thedevelopnik
+
+## What
+
+Implement support for SPIFFE Federation within Teleport Workload Identity,
+including support for a SPIFFE based join method based on the federation
+mechanism.
+
+See https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md
+for the SPIFFE Federation specification. This will be referred to as
+"the specification" throughout this document.
+
+This feature will be gated behind a Teleport Enterprise license.
+
+## Why
+
+Implementing support for SPIFFE Federation has the following benefits:
+
+- Improves the feature parity of Workload Identity with competing solutions,
+  increasing appeal to prospects.
+- Allows customers to adopt Teleport Workload Identity gradually, by federating
+  with existing SPIFFE based systems, rather than needing to perform a complete
+  migration.
+- Supports an Active/Active and Active/Passive high availability model.
+- Support segmented environments using multiple Teleport clusters which
+  federate with one another.
+- Enables the implementation of a SPIFFE-based join method.
+
+Implementing support for a SPIFFE join method:
+
+- Secure alternative to the `token` based join-method for OnPrem environments.
+- Allows for more advanced join implementations based on Teleport Workload 
+  Identity, e.g distributing SPIFFE SVIDs using an out of band mechanism and
+  then joining using the SPIFFE SVIDs.
+
+## Implementation
+
+### Federation
+
+SPIFFE federation relationships are one-way. One trust domain can be configured
+to trust identities issued by another trust domain, but the reverse may not 
```

Comment:
Hah - I think technically this is already the case today as we currently just pull the CAs, so if you're in a leaf cluster, the root cluster SPIFFE CA is already there. But, we should definitely formalize that as a thing that happens. I think it should be possible to make the federation syncer automatically create a SPIFFEFederation resource in these cases as well so it's "visible" in the same way. I'll have a think about it.

Reply:
So I've opted against this for now. I think there's a lot of cases where folks will want this relationship to be "two way" and I think providing an easier mechanism for configuring that (or the one way relationship) makes more sense rather than having a default that won't actually serve most people's needs. We'll achieve that through a custom "bundle source" that's optimised for Teleport clusters and will leverage the existing CA syncing that Teleport performs. It'll make it easier to setup, but leave the control in the hands of the operator.